# Donovan Feist, u09249825, CH EN 2450

## Homework 2

### Problem 2 modified assignment

Write a Python script that computes the natural logarithm function using a Taylor series
expansion up to a desired number of significant digits.

In [41]:
def natural_log_of_1_minus_x(x,m):
    #x is the value being evaluated by the function; m is the number of significant digints required
    a=0.5*10**(2-m) #a is the maximum value for the approximate relative error (in %) that still guarantees m accurate significant digits
    ϵa=10**6 #pick some value that will definite be larger than a at first
    oldval=0
    terms=1 #counter that keeps track of how many terms are in the taylor expansion
    while ϵa > a:
        newval=oldval+(-1)*(x**terms)/terms #taylor series for ln(1-x)
        ϵa=abs((newval-oldval)/newval)*100 #approximate relative error (in %)
        oldval=newval
        terms+=1
    return terms-1,newval

In [42]:
sig=[1,2,4,6,12] #the user can enter the desired number of significant digits here
x=-0.1 #here the user can define the value of x to be evaluated by the natural log function (x must be between -1 and 1)

print("# of terms      value")
print("----------------------------")
for m in sig:
    terms, newval=natural_log_of_1_minus_x(x,m)
    print(terms,"             ", newval)

# of terms      value
----------------------------
3               0.09533333333333334
3               0.09533333333333334
5               0.09531033333333334
7               0.09531018095238097
13               0.09531017980432552


In calculating $ln(1.1)$, the following values were obtained:

| # of terms | significant digits | value           |
|------------|--------------------|-----------------|
| 3          | 1                  | 0.09            |
| 3          | 2                  | 0.095           |
| 5          | 4                  | 0.09531         |
| 7          | 6                  | 0.0953102       |
| 13         | 12                 | 0.0953101798043 |